x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/index/EsIndex.java

```diff
@@ -19,18 +19,38 @@
 
 import static java.util.stream.Collectors.toMap;
 
-public record EsIndex(String name, Map<String, EsField> mapping, Map<String, IndexMode> indexNameWithModes) implements Writeable {
+public record EsIndex(
+    String name,
+    Map<String, EsField> mapping,
+    Map<String, IndexMode> indexNameWithModes,
+    /** Fields mapped only in some (but *not* all) indices. Since this is only used by the analyzer, it is not serialized. */
+    Set<String> partiallyUnmappedFields
```

Comment:
mark the field transient to signal it's not going to be serialized.
Additionally instead of returning the set, define a method that checks the present  - this makes it easier in the future to add custom matching directly in this class.